src/flamegraph/svg.rs

```diff
@@ -48,224 +48,270 @@ pub(super) struct StyleOptions<'a> {
-    svg.write_event(Event::CData(BytesText::from_escaped_str(include_str!(
-        "flamegraph.js"
-    ))))?;
-    svg.write_event(Event::End(BytesEnd::borrowed(b"script")))?;
+    pub(super) fn write_prelude<'a, W>(
+        self: &SVGWriter,
+        svg: &mut Writer<W>,
+        style_options: &StyleOptions<'a>,
+    ) -> quick_xml::Result<()>
+    where
+        W: Write,
+    {
+        svg.write_event(Event::Start(BytesStart::borrowed_name(b"defs")))?;
+        svg.write_event(Event::Start(BytesStart::borrowed(
+            br#"linearGradient id="background" y1="0" y2="1" x1="0" x2="0""#,
+            "linearGradient".len(),
+        )))?;
+        svg.write_event(Event::Empty(
+            BytesStart::borrowed_name(b"stop").with_attributes(
+                iter::once(("stop-color", &*style_options.bgcolor1))
+                    .chain(iter::once(("offset", "5%"))),
+            ),
+        ))?;
+        svg.write_event(Event::Empty(
+            BytesStart::borrowed_name(b"stop").with_attributes(
+                iter::once(("stop-color", &*style_options.bgcolor2))
+                    .chain(iter::once(("offset", "95%"))),
+            ),
+        ))?;
+        svg.write_event(Event::End(BytesEnd::borrowed(b"linearGradient")))?;
+        svg.write_event(Event::End(BytesEnd::borrowed(b"defs")))?;
 
-    svg.write_event(Event::Empty(
-        BytesStart::borrowed_name(b"rect").with_attributes(vec![
-            ("x", "0"),
-            ("y", "0"),
-            ("width", &*format!("{}", super::IMAGEWIDTH)),
-            ("height", &*format!("{}", style_options.imageheight)),
-            ("fill", "url(#background)"),
-        ]),
-    ))?;
+        svg.write_event(Event::Start(
+            BytesStart::borrowed_name(b"style").with_attributes(iter::once(("type", "text/css"))),
+        ))?;
+        svg.write_event(Event::Text(BytesText::from_plain_str(
+            ".func_g:hover { stroke:black; stroke-width:0.5; cursor:pointer; }",
+        )))?;
+        svg.write_event(Event::End(BytesEnd::borrowed(b"style")))?;
 
-    // We don't care too much about allocating just for the prelude
-    let mut buf = StrStack::new();
-    write_str(
-        svg,
-        &mut buf,
-        TextItem {
-            color: "rgb(0, 0, 0)",
-            size: super::FONTSIZE + 5,
-            x: (super::IMAGEWIDTH / 2) as f64,
-            y: (super::FONTSIZE * 2) as f64,
-            text: (&*opt.title).into(),
-            location: Some("middle"),
-            extra: None,
-        },
-    )?;
+        svg.write_event(Event::Start(
+            BytesStart::borrowed_name(b"script")
+                .with_attributes(iter::once(("type", "text/ecmascript"))),
+        ))?;
+        svg.write_event(Event::CData(BytesText::from_escaped_str(&format!(
+            "\
+    var nametype = '{}';
+    var fontsize = {};
+    var fontwidth = {};
+    var xpad = {};
+    var inverted = {};
+    var searchcolor = 'rgb(230,0,230)';",
+            self.name_type,
```

Comment:
Here we probably have to do some escaping _just_ in case the font name has a `'` in it!

Reply:
Do we have a valid font with `'`?
Maybe it is better to validate it in arguments and to fail fast?

Reply:
I don't _know_ of one, but that doesn't mean there aren't any :p
I don't think there's a good reason do _disallow_ such fonts? Probably better to just do
```rust
self.name_type.replace('\'', r#"\'"#)
```

Reply:
Also, I just realized, this isn't the font name, but rather the "name" assigned to functions (basically the prefix `Function:`). There are _definitely_ cases where the user might use `'` in those.

Reply:
I think we need to have tests for this. It is already escaped in xml, so it should be escaped properly, probably replace is not a good option either.

Reply:
Ah, I'm not worried about the XML, but rather about the fact that we're creating a `'`-enclosed JavaScript string directly. It should be the case that `'` is the only character we _need_ to escape, though `\` may also be a good candidate (specifically by making it `\\`).